### DISCUSSION TITLE
future of rnmapbox and the maps repo

### DISCUSSION THREAD

**Discussion by ferdicus** [rnmapbox/maps#1680] [General]
__Quick update of state in 2023 by @mfazekas__: 
- V10 implementation is on the way for release, getting more and more stable
- MapLibre is deprecated, but they have a fork
- Code has been migrated to typescript

Discussion ticket on the developments around `Mapbox v10`/ `Mapbox "old"`/ `MapLibre`  and where _this_ repo is going in the future.

If you haven't already, you should read [`@systemlevels`](https://github.com/systemlevel) [reply](https://github.com/rnmapbox/maps/pull/1222#issuecomment-996221674) to [this ticket](https://github.com/rnmapbox/maps/pull/1222).

When commenting, I'd like you to keep some things in mind:
* there are just a few regular contributors for this project
* they are spread thin, as they need to:
  support multiple SDK integrations (Mapbox pre/ post v6 | MapLibre)
  new influx of EXPO users with their share of support queries
  fix bugs
  review PRs 
  keep `/example` up to date with RN/ Android/ iOS versions
* `Mapbox "old"` & `MapLibre` codebase are diverging from each other more and more
* adding a third stream to this mix would maybe be a bit too messy
* how many users would actually jump to `Mapbox v10` once it would be fully supported for RN
* how many are using `MapLibre` vs `Mapbox`

In any case, this is just a playground to discuss these sorts of things.

Additionally I'd like to let everyone know, that I'll be stepping away from being a maintainer in the near future (this has been in the making for a while, @mfazekas and I talked about this) as I'm moving on to another project internally.

Nevertheless, our company is still heavily in support of all things RN/ Mapbox as we are using it as a vital component in one of our apps. During my transition phase there probably won't be too much involvement, however I've been told, that mid of '22 engineering time will be freed to support this (follow-up repo?) again.

**Comment by maxammann** [comment]
We are using the MapLibre backend for our open-source app [Integreat](https://github.com/digitalfabrik/integreat-app). Because we target to launch in the open-source app market F-Droid we benefit from the libre fork.
We also host all of our map data and tiles on our own infrastructure. Essentially there is no need for us to use the mapbox/maptiler API. Hosing tiles is relatively easy and cheap.

So for us there are no benefits with upgrading to Mapbox V10. We would not benefit from the new features.

I'm not sure what the feature-wise goals of MapLibre are. But I doubt that it will diverge from `Mapbox "old"`. The last commit in the maplibre project was almost a month ago: https://github.com/maplibre/maplibre-gl-native
I think they are only fixing bugs and providing infrastructure like package hosting or CI.

**Comment by sheppard** [comment]
I too am focused on the MapLibre use case at the moment.  I assume that at some point (potentially within the next year) everyone using Mapbox "old" is going to have to switch to either MapLibre or Mapbox V10.  Given that the respective code bases will continue to diverge, it likely makes the most sense to split into two repositories.  I don't think anyone should feel like they are obligated to help maintain both versions.  In the short term that means there will be even fewer contributors to each repository, but a narrower focus will hopefully make it easier to bring new contributors in.  (No one wants to have their PR rejected because it doesn't work with an SDK they aren't using.)

While there are a few ways to do this, here's a recommendation to start from:

 * This repository becomes the Mapbox V10 repository.  Mapbox "old" support can be provided through a 8.x maintenance branch temporarily, then dropped entirely once enough users have upgraded (or maintenance becomes too much of a burden).  The Github org and NPM scopes can change if needed to reflect the technology underlying V10.
 * A new repository is created under @maplibre for maximum visibility with that community (assuming they agree).  The new repo could be a fork of this one, but it might be cleaner to re-import everything (like the MapLibre core libraries have done).  The repository could be published on npm as @maplibre/react-native-maps or something similar.

I'm not able to commit to much help in the short term, but I will try to pitch in as I am able.

**Reply by maxammann** [reply]
I also already thought about messaging @maplibre. I think this is a good idea!

**Reply by wipfli** [reply]
If you want to have a repo in the MapLibre GitHub organization, feel free to open a "New Repository Checklist" issue at https://github.com/maplibre/maplibre/issues/new/choose and then we can all discuss it there.

**Reply by wipfli** [reply]
@lseelenbinder started the process creating a fork in the MapLibre organization. See https://github.com/maplibre/maplibre/issues/134

**Comment by Dakuan** [comment]
We are mapbox customers and have built our business on React Native while in early stages. We really dont want to be getting involved with running our own tile servers or anything like that. Lack of support for v10 would see us stop being mapbox customers and go back to using native maps, which would be a shame. We're already pretty troubled by how little interest mapbox are paying to this issue.

**Reply by kneubuehl** [reply]
We are in the same camp as @Dakuan and others on this discussion. We'd really like to see further support for this repo and for V10. A core function of our app is built on this.

Does anyone have direct contact with MapBox and/or their view on React Native? How can we get them more engaged.

**Reply by systemlevel** [reply]
Yes, we have been in contact with them and there's some initial movement on that as well.

**Reply by kneubuehl** [reply]
Good to hear that @systemlevel, if there's anything we can do to voice support or add pressure on them, please let me know. I have voiced this with some of their product teams (during a feedback session) as well as on Twitter, but they have not directly responded or addressed it.

**Comment by danidaryaweesh** [comment]
We are also Mapbox customers and rely heavily on this React Native repository. Supporting V10 for React Native would really be great and I think that once it is stable then it makes sense to stop supporting V6 on this repository so it would be easier for maintenance as well. I agree with @sheppard 👍🏻

**Reply by bencallis1** [reply]
100% agree with @danidaryaweesh

**Reply by akougnon7** [reply]
why can we not able to use MapboxGL.UserLocation without this error ?
could not invoke rctmgllocationmodule.start

**Comment by Gnative** [comment]
I am very interested in a Mapbox v10 version of this repo as my app uses heavily Mapbox services. If I was in a financial position to contribute to this development I would have and will in the future if it becomes possible, along with development time.

**Comment by schiller-manuel** [comment]
Would love to use V10 in the future.

**Comment by bjarn** [comment]
Same here, would love to use V10 in the future.

Currently using the "old" Mapbox. When we got blocked by the M1 issue where Mapbox does not compile on, we tried to use MapLibre. However, after trying some things out, we still preferred Mapbox.

As I'm quite new still to RN, its libraries in particular, I am hesitant to contribute, but am willing to do that in the future.

**Comment by jeremyvinec** [comment]
I am looking forward to the V10 version.
I really like mapbox and these maps! and I would like to use it for 3D terrain and improve the performance of my cards.

**Comment by ansarikhurshid786** [comment]
I also use this library for many projects. I am willing to support

**Reply by timkuilman** [reply]
At Chargetrip we are using Mapbox for an EV navigation application. We are currently implementing RN Mapbox Navigation which shares dependencies with this project. This prevents us from updating the underlying Navigation SDK's (1.4.2 is the last 1x version, latest version is 2.1.1). In order to use the latest navigation SDK (which uses V10 for maps) we need this project to be updated to the latest SDK. 

@systemlevel @ferdicus @mfazekas and all others involved: keep up the great work!

**Reply by danidaryaweesh** [reply]
We are using this library for Turn by Turn navigation and extended the library with our own logic and code. @timkuilman How well does RN Mapbox Navigation work?

**Reply by timkuilman** [reply]
It works but we had to modify it quite extensively. We are not using the Mapbox instructions and instead use our own EV-optimised route instructions. We modified the RN component to accept the instructions as a JSON. 

Other than that the RN navigation project is behind on the v1.x releases. For iOS it uses the 1.3.0 SDK (instead of 1.4.2) and on Android it's on 1.5.0 instead of 1.6.2. We're updating that on our fork. In order to jump to the 2.x versions we need the updated react-native-mapbox-gl. 

Little bit off topic sorry, DM if you have more questions. :)

**Comment by ludvigeriksson** [comment]
I just started my first React Native project in which a map has the main focus so it is heavily dependant on this library and Mapbox. I would love to see v10 support. Since I'm totally new to both React Native & Mapbox, I'm not sure I have much knowledge to contribute with, but would maybe be able to help with smaller tasks. Maybe you can consider a Patreon (or OpenCollective or similar) so that multiple smaller users can pitch in financially?

**Comment by d3vhound** [comment]
Would love to see v10 support. I know 2 people willing to sponsor if movement is made to upgrade to v10.

**Reply by systemlevel** [reply]
Awesome, thank you.  We are moving on it. We'll be in touch.

**Reply by ferdicus** [reply]
@systemlevel, did you already consider cashing in all these statements of sponsoring?
At the moment you guys are paying this our of pocket, right?

**Reply by systemlevel** [reply]
That's correct.  We are the sole sponsor at this time.  I've been taking note of those interested and plan on coordinating that sponsorship effort once we get this landed into the community.

**Reply by ferdicus** [reply]
let people put their money where their mouth is... that sounded weird, you know what I mean :P

**Comment by haron68** [comment]
I work for a company that depends on this library would love to contribute and sponsor as well. We would love to see support for Mapbox v10. I haven't tried out MapLibre and I'm not sure how popular it is if it's worth supporting

**Reply by Gnative** [reply]
🤘

**Comment by tomgreco** [comment]
This library is the main package we use in our app, I'd be happy to assist where possible in maintaining this.

**Comment by Alexandria** [comment]
Question what's the process for issues that are stale and could probably be closed? [This one for example](https://github.com/react-native-mapbox-gl/maps/issues/1467)

Do we tag someone? Do they stay open until the person who logs it closes it?

**Comment by mfazekas** [comment]
@Alexandria  I'd vote for closing it. We have way to many issues anyway.

**Reply by Alexandria** [reply]
I hear ya. I don't have the permissions to close this/ assign the issue to myself.   But I do still plan to spend an hour or so every Friday to help y'all out with pruning the issue backlog.  Just let me know the best way to flag these and I'll get it done 🙂

**Comment by mfazekas** [comment]
We're planning on renaming the GitHub organization from `react-native-mapbox-gl` to ~`react-native-mapbox`~ `rnmapbox`. Anyone has objection against it?!

**Reply by haron68** [reply]
If it doesn't cause any trouble with npm or yarn I think that's a more appropriate name

**Reply by ansarikhurshid786** [reply]
There is 2 libraries mapbox and mapbox-gl. mapbox-gl has 3d support as well, so i this react-native-mapbox-gl is perfect name.

**Reply by mfazekas** [reply]
@ansarikhurshid786 note that with v10 mapbox also dropped ‘gl’ from their map libs names. For example: https://github.com/mapbox/mapbox-maps-ios/ as they are no longer OpenGL based on all platform

**Reply by BelyaninD** [reply]
@mfazekas will you support maplibre in the future?

**Reply by naftalibeder** [reply]
That makes a lot of sense! Regarding what @BelyaninD asked, since the name does influence the tone of the project, it is worth deciding whether this would be all-in on Mapbox v10 (no Maplibre support).

But if so, that would be really clean/appropriate rename.

**Reply by mfazekas** [reply]
On `maplibre` support, the not (yet) public v10 branch still supports MapLibre/old Mapbox libraries. 

The way it's implemented is that the native part has a v10 version and current version. And JavaScript code has very minimal `if`-s here and there. 

I assume this will work in short term, but in long run, forking the project might be more optimal.

**Reply by mfazekas** [reply]
Update: It seems that `react-native-mapbox` org/scope is already claimed on npm. So we'll go with `rnmapbox`

**Comment by bencallis1** [comment]
I'm a founder of a new tech company that's heavily focused on real estate. I'd like to develop a plan for getting your team the financial support needed to keep this repo progressing. Would you mind sending me a DM to further discuss this?

**Comment by JorgeQuevedoC** [comment]
I have handle to update a turn by turn navigation library to latest v2.X.X version but there are some incompatibilities that v10 would solve. If I can help to get this version ready please let me know

**Reply by JorgeQuevedoC** [reply]
Hi @systemlevel according to your reply about the 75% - 80% of progress you had by December is there an estimated time for the changes to be released?

**Reply by systemlevel** [reply]
We don't have a specific date yet but I can give a status update.  We are working to get the lib up on NPM and also working out some initial issues.  I think we are getting pretty close to getting it out into the open-source community.  

@bencallis1 would love to talk further.

**Comment by babeone** [comment]
This repo is gold, hope to see a v10 support.

**Comment by maxammann** [comment]
I think the consensus of this discussion is that the majority aims to use the new version of mapbox. I would like to raise the awareness in this thread about the license changes introduced by mapbox.

Since V10 mapbox is no longer free and open-source software, while `react-native-mapbox-gl` continues to do so. I would like to highlight that people are using their free and spare time to contribute to `react-native-mapbox-gl` without receiving anything in return. This essentially means that maintainers of this project offer free work-force to mapbox.
**Therefore, I would suggest to approach mapbox and put pressure on them to support this project. Support could happen through infrastructure, paid maintainers or symbolical support (put a label on this project: officially supported by mapbox).**

Another issue in the future could be that the project could get into legal troubles. Reading and understanding the source code of mapbox is no longer possible like before. I would not say that mapbox would cause any real legal issues, but there always is the chance.

I kind of see an ethical problem here: The open-source community supports mapbox, a business which caused an ecosystem-split by going partly non-free. I think the continue should decide early in which direction we want to continue.


This thread is neither pro or contra supporting V10 or not. I just want to highlight that things have changed since the license change. Open-source maintainers should be aware that the mapbox ecosystem is non-free.

**Reply by Gnative** [reply]
Nice @maxammann 👌. It's probibly a good idea to make sure this is clearly stated in the intro README.md.

If this package didn't exist we would have to be building our own implementation of Mapbox for RN if we wanted to use v10 features of Mapbox 3D maps etc.. I assume that a lot us are still using Mapbox services and are aware of the license and recent history with Mapbox. 

I hope that Mapbox will come to the table a little to help the awesome devs putting in the time to update this package..  

Looking forward to seeing the initial development and being able, hopefully, to contribute where I can.

**Comment by mfazekas** [comment]
We've renamed the organisation to `rnmapbox` we'll also migrate the npmjs organization to `@rnmapbox`.

The v10 branch is now public as well - see #1750

**Reply by naftalibeder** [reply]
Congratulations and thank you to everybody involved! This is such a great contribution to the community.

**Comment by mfazekas** [comment]
I plan to do these changes in the near future:
- [x] publish a 8.6.0-beta.0 npm beta release from current master
- [x] make a [v8 branch](https://github.com/rnmapbox/maps/tree/v8) from current master (in case we need to publish smaller updates to npm 8.6.* versions) 
- [x] change default branch from `master` to `main`
- [x] merge the v10 branch, which means:
  - [x] supports v10 implementation (requires setup/config)
  - [x] default to maplibre (as mapbox v10 requires setup steps anyway tokens in .netrc, etc.)
  - [x] pre v10 mapbox libs might or might not work
  - [x] both v10 and maplibre will run on CI

After that:
- I'm sure pre v10 mapbox lib support will be dropped
- MapLibre support might go to another branch/repo, we'll see

**Reply by naftalibeder** [reply]
~Just want to note that we've done a test of the v10 branch on iOS, and the map does not display. (See https://github.com/rnmapbox/maps/issues/1804.) If anyone has any ideas, I'd love to hear them.~ This was simple user error. Disregard!

**Comment by waymond91** [comment]
🔥 🔥 🔥 Loved the previous version and am happy the project is still alive. 🔥 🔥 🔥 

Embarassingly I'm one of the "expo influx users" - I'm leaning really heavily on the previous version of this repo (react-native-mapbox-gl) for my biggest client project. 

Not sure how I can help but will be happy to if I find a pathway.
My clients are also very excited for 3d assets and terrain features - would keep money flowing and keep food on the table

**Comment by kneubuehl** [comment]
Hi all, this morning I received an email from MapBox newsletters asking about our upgrade to V10. I replied to that email mentioning that there is a large community here using MapBox for react native, that wants their support and engagement for the project. We've built some central pieces of our app around functionality that depends on MapBox and we'd like to see this project continue and move forward with their native products.

In order to help raise that awareness, I ask that everyone who also receives that newsletter to please respond asking for their engagement and support of the community. Even if you didn't receive that email, send them one at support@mapbox.com and let them know of your interest.

If any of you have news or updates from MapBox regarding their strategy or approach for react native, please state it here as well. Hopefully we can start raising more awareness and get their support again for this community.

**Reply by alaurenz** [reply]
Thank you for alerting us, I just sent an email to [support@mapbox.com](mailto:support@mapbox.com) about our interest in supporting.

**Comment by bilal-korir** [comment]
I've been using ReactNative Mapbox for the last two years. I add my voice to the community here regarding V10. Yes, I would like to test and use V10.

Thank you very much for all your work keeping this repository up to date :)

**Comment by saha-chan** [comment]
I've been using ReactNative Mapbox for quite a while and loved it. My atmost respect to all developers behind making of such an amazing project. Yes, I would love to use V10 for my RN projects.

**Comment by 1mike12** [comment]
Can you add a donation link or something? I think it's safe to say a lot of us are using this in commercial ways, and we're gonna be paying mapbox anyways, it should be fair to send resources to the project that is doing the work that mapbox should've done themselves

**Comment by bencallis1** [comment]
I fully agree and would love to donate

BEN CALLIS
CEO @IVYHOUSE
801.368.9442

Sent via Superhuman iOS ***@***.***>

**Comment by learner-long-life** [comment]
I would also like to support this community with funds or developer time, and have written to support@mapbox.com .
Let me know how else I can help.

**Comment by mattijsf** [comment]
I personally would like to see support of maplibre move to another repo and keep focus here on being a wrapper for the official platform sdk's from mapbox. I think in the end it will be too distracting & confusing when the two sdk's deviate from each other in terms of features / bugs / api's.

**Reply by mfazekas** [reply]
I agree, hopefully we can start doing releases of 10.0 from #main in the summer. After that we'll remove maplibre together with old mapbox-gl, and #main will be just the v10 version of mapbox library.

**Reply by Rakiah** [reply]
So essentially, you're dropping support for MapLibre entirely ?

**Reply by mfazekas** [reply]
@Rakiah yeah, we can fork the repo to @rnmapbox/maplibre.
We haven't received contributions for MapLibre. MapLibre is a different native codebase and it's more and more pain to maintain, and most of our contributors are interested and testing only in v10.

**Reply by itekhi** [reply]
@mattijsf I was trying something right now, and when installing through expo I didn't even realize that there's 2 ways of `app.json` plugin config, `maplibre` and `mapbox`, it was really confusing, so I agree with distracting & confusing 😄

**Comment by ricardodolnl** [comment]
So what's the status of 10.0 release at this moment? Any idea when we can expect the official release?

**Reply by mfazekas** [reply]
@RicardoDolNL we're now publishing beta releases to npm.

**Comment by ansarikhurshid786** [comment]
I am also heavily depend on this library for all my current projects. I am planning to support as much as i can.

**Comment by nzjs** [comment]
Hello, thank you for all of the hard work with the migration to a stable V10 so far. :) 

I received the below email from Mapbox at the start of this month and wondered what the current status is with V10 release? I see there are recent PRs for this work so that looks like good news! 

What can we do to help? Should we be emailing [support@mapbox.com](mailto:support@mapbox.com) again to ask for their support on this project, or does anyone here have direct contact with the SDK team?

And finally, is it okay to use the beta releases in the interim until the stable V10 is ready?

Full email from Mapbox:

> Our records indicate that you’re using an older version of the Mapbox Maps SDK. Following up on our communication in April 2022, we’re writing to give you one final reminder that we’ll be ending support for the software you’re using on April 4, 2023 - and to ask how we can support your upgrade to the current version of our SDK: Mapbox Maps v10.
> 
> We released v10 in October 2021 and have been delighted by the response it’s received. Customers who’ve already upgraded are benefiting from an SDK that’s more performant, easier to work with, and which brings new capabilities like 3D terrain, globe view, and faster offline downloads. [You can learn more about v10's advantages here.] Or, to get a better sense of what upgrading will require, have a look at our migration guides ([iOS], [Android].
> 
> **Pre-v10 versions of our SDK may continue to work past April 4, 2023,** but that is the date at which we will stop publishing bug and compatibility fixes. It’s also the point beyond which we will no longer be able to guarantee support or compatibility.
> 
> We realize that your app might be well-served by the SDK version you’re currently using. The move to v10’s modern foundations will deliver real benefits to you and your users, and its modernized architecture will mean we can continue delivering the best mobile maps experience possible.
> 
> We’re eager to help you with this transition. Have a look at the migration guides, and if you need assistance or have any questions, please reach out to your Account Manager or our [Support] team.
> 
> Kind regards,
> Maps SDK Team

**Reply by mfazekas** [reply]
@nzjs we've released rc0 from 10.0 and we plan to release final ideally in spring. But feel free to use beta releases in production, if it works for you.

**Reply by mfazekas** [reply]
FWIW I've just released the final release of 10.0

**Reply by Gnative** [reply]
@mfazekas awesome work.. thank you.

**Reply by tomskopek** [reply]
I want to echo my thanks @mfazekas and other contributors
We bumped to mapbox 10 this week and it was quite painless
🙏

**Reply by veb-ioki** [reply]
Thank you so much for your endurance and all  effort you put into this @mfazekas!
🙏

**Comment by KiwiKilian** [comment]
Thanks everybody who contributed to v10, especially @mfazekas! 🚀
